src/ints/int10_modes.cpp

```diff
@@ -1953,44 +1956,221 @@ static cga_colors_t handle_cga_colors_prefs_ibm5153(const std::vector<std::strin
+				log_warning("RGB-triplet values must contain only digits");
+				return -1;
+			}
+
+			int32_t value;
+			if (!sscanf(c.c_str(), "%d", &value)) {
+				log_warning("could not parse RGB-triplet value");
+				return -1;
+			}
+			if (value > 0xff) {
+				log_warning("RGB-triplet values must be between 0 and 255");
+				return -1;
+			}
+			return value;
+		};
+
+		const auto r = parse_component(r_string);
+		const auto g = parse_component(g_string);
+		const auto b = parse_component(b_string);
+
+		if (r < 0 || g < 0 || b < 0)
+			return {};
+		else
+			return RGBEntry{static_cast<uint8_t>(r),
+			                static_cast<uint8_t>(g),
+			                static_cast<uint8_t>(b)};
+	}
+	default:
+		log_warning("colors must be specified as 3 or 6 digit hex values "
+		            "(e.g. #f00, #ff0000 for full red) or decimal RGB-triplets "
+		            "(e.g. (255, 0, 255) for magenta)");
+		return {};
+	}
+}
+
+// Parse `cga_colors` using a standard parsing approach:
+// 1. first tokenize the input into individual string tokens, one for each
+// color definition
+// 2. validate and parse the color tokens
+std::optional<cga_colors_t> parse_cga_colors(const std::string &cga_colors_prefs)
+{
+	const auto tokens = tokenize_cga_colors_pref(cga_colors_prefs);
+
+	if (tokens.size() != num_cga_colors) {
+		LOG_WARNING("INT10: Invalid 'cga_colors' value: 16 colors must be specified "
+				    "(found only %u)", static_cast<uint32_t>(tokens.size()));
+		return {};
 	}
 
 	cga_colors_t cga_colors = {};
+	bool found_errors = false;
 
-	for (size_t i = 0; i < cga_colors_prefs.size(); ++i) {
-		auto color = cga_colors_prefs[i];
-
-		if (uint32_t c; (color.size() == 6) && sscanf(color.c_str(), "%x", &c)) {
-			auto to_uint8 = [](const int c) {
-				constexpr auto col_min = 0;
-				constexpr auto col_max = 0x3f;
-				return static_cast<uint8_t>(
-				        clamp(c, col_min, col_max));
-			};
-
-			const auto red   = to_uint8(c >> 16 & 0xff);
-			const auto green = to_uint8(c >> 8 & 0xff);
-			const auto blue  = to_uint8(c & 0xff);
-			cga_colors[i]    = {red, green, blue};
+	for (size_t i = 0; i < tokens.size(); ++i) {
+		const auto token       = tokens[i];
+		const auto maybe_color = parse_color_token(token,
+		                                           static_cast<uint8_t>(i));
+
+		if (maybe_color.has_value()) {
+			auto color = maybe_color.value();
+
+			// For now we only support 18-bit colours (6-bit
+			// components) when redefining CGA colors. There's not
+			// too much to be gained by adding full 24-bit support,
+			// and it would complicate the implementation a lot.
+			color.red   >>= 2;
+			color.green >>= 2;
+			color.blue  >>= 2;
+
+			cga_colors[i] = color;
 		} else {
-			LOG_WARNING("INT10: Error parsing 'cga_colors': invalid value '%s' at index %u, "
-			            "using default CGA colors",
-			            color.c_str(),
-			            static_cast<uint32_t>(i));
-			return cga_colors_default;
+			found_errors = true;
```

Comment:
PVS studio is flagging this:

![2022-08-14_10-05](https://user-images.githubusercontent.com/1557255/184547760-96f92869-055d-4dc7-a72b-372f48952bd6.png)

The PVS folks seems to feel that if we've already called `had_value()`, then we shouldn't need to use `.value()`.

Here's their explanation: https://pvs-studio.com/en/docs/warnings/v830/

It looks like `std::optional` is a nice managed pointer, instead of having to return a raw pointer, but it's use can be quite similar (in that we can bool-evaluate directly, and use -> and * to dereference it).

If we treat the return value like a pointer, we could do something like:

```c++
if (auto color = parse_color_token(token, static_cast<uint8_t>(i)); !color) {
	found_errors = true;
} else {
	// For now we only support 18-bit colours (6-bit
	// components) when redefining CGA colors. There's not
	// too much to be gained by adding full 24-bit support,
	// and it would complicate the implementation a lot.
	color->red >>= 2;
	color->green >>= 2;
	color->blue >>= 2;

	cga_colors[i] = *color;
	}
}
```

Or it could be flipped to the positive-first and the `found_errors` put in the else/bottom.

Reply:
Oh nice, wasn't aware it can be used like this. Changing to your solution.